### ISSUE TITLE
io_uring is slower than epoll

### ISSUE THREAD

**Issue by ghost** [axboe/liburing#189]
EDIT: I have made available detailed benchmark with epoll that shows this in a reliable way: https://github.com/alexhultman/io_uring_epoll_benchmark

---

Don't get me wrong, since I heard about io_uring I've been all about trying to scientifically reproduce any claim that it improves performance (theoretically it should, since you can have fewer syscalls and I like that idea). I would love to add support for io_uring in my projects, but I won't touch it until someone can scientifically show me proofs of it outperforming epoll (significantly).

I've tested three different examples claiming to be faster than epoll, tested on Linux 5.7 without spectre mitigations, tested on Linux 5.8 with spectre mitigations. Tested Clear Linux, tested Fedora. All tests points towards epoll being faster for my kind of test.

What I have is a simple TCP echo server echoing small chunks of data for 40-400 clients. In all tests epoll is performing better by a measurable amount. In no single test have I seen io_uring beat epoll in this kind of test.

Some io_uring examples are horribly slow, while others are quite close to epoll perforamance. The closest I have seen it go is this one: https://github.com/frevib/io_uring-echo-server comparing to this one: https://github.com/frevib/epoll-echo-server

I have tested on two separate and different machines, both with the same outcome: epoll wins.

Can someone enlighten me on how I can get io_uring to outperform epoll in this case?

**Comment by alexhultman** [comment]
https://github.com/frevib/io_uring-echo-server/issues/8

**Comment by isilence** [comment]
That's definitely a great thing to do. I don't know whether anybody have time for that at the moment, though.
io_uring-echo-server looks much bulkier from the last time I've seen it.

**Comment by grober82** [comment]
Maybe your benchmark is full of errors

**Comment by alexhultman** [comment]
> Maybe your benchmark is full of errors

It's not my benchmark and the benchmark I reference is the one @axboe himself has referenced on Twitter, showing that he interprets it as true. That's why I would like @axboe himself to write a benchmark without issues so that we can actually prove this thing works better.

**Comment by markpapadakis** [comment]
For what it’s worth there is a whole lot more to it than trivial one to one comparison. 
- you save many syscalls that would be required to modify the epolll kernel state (adding, removing monitored FDs, updating per FD state etc). This becomes extremely important as the number of managed FDs increases - exponentially so. 
- you can register FDs (eg listener FDs and long lived connections FDs) which means you dont have to incur the cost of kernel looking up the file struct and checking for access every time. 
- and, obviously, you don’t get to monitor just network sockets. You can do so much more. 



@markpapadakis

**Comment by Qix-** [comment]
@markpapadakis Have you taken a look at the benchmarks that are quoted above? They use many connections at once, thus many FDs at once. uring still consistently performs worse, in conditions that mimick production usage well enough without actually being production.

Part of the scentific process is to be able to reproduce claims like those being made (60%+ increase in performance over epoll, apparently that was 99% at one point but bugs were found). 

Even if this were a 5% increase, I'd be for it. However, I simply do not understand where these results are coming from, and everyone on twitter seems to be more interested in patting themselves on the back rather than addressing criticisms of things (sorry if that sounds harsh).

Outrageous claims require outrageous evidence, especially when it comes to claims that could completely renovate the space.

**Comment by alexhultman-2fa** [comment]
@markpapadakis We all know the theory. It is not that hard to understand, quite basic actually.

But theory means nothing if actual reality mismatches with theorized conclusions (any scientist ever).

I would be happy and excited if io_uring could improve performances but so far no benchmark can show this.

All I'm asking for is scientific proofs. I am a scientist, not a believer.

**Comment by alexhultman-2fa** [comment]
I think it is time to take this to the Linux kernel mailing list since @axboe has ignored this criticism entirely.

**Comment by Qix-** [comment]
@alexhultman Check netty/netty#10622 - I won't have any bandwidth for the next day or two, but maybe that will pique your interest.

**Comment by grober82** [comment]
I think that @axboe knows that uring is slower. Otherwise he would answer.

**Comment by axboe** [comment]
I have not ignored any of this, but I've been way too busy on other items. And frankly, the way the tone has shifted in here, it's not really providing much impetus to engage. I did run the echo server benchmarks back then, and it did reproduce for me. Haven't looked at it since, it's not like I daily run echo server benchmarks.

So in short, I'm of course interested in cases where io_uring isn't performing to its full potential, it most certainly should. I have a feeling that #215 might explain some of these. Once I get myself out from under the pressure I'm at now, I'll re-run the benchmarks myself.

**Comment by romange** [comment]
I can testify that io_uring is much faster than epoll. Please use kernel 5.7.15 for your benchmarks. 
This server https://github.com/romange/gaia/tree/master/examples/pingserver reaches 3M qps on a single instance for redis-benchmark (ping_inline API) on c5n ec2 instances.

**Comment by alexhultman** [comment]
I have made available a new, detailed benchmark that shows io_uring is reliably slower than epoll:

https://github.com/alexhultman/io_uring_epoll_benchmark

> I can testify that io_uring is much faster than epoll. Please use kernel 5.7.15 for your benchmarks.
This server https://github.com/romange/gaia/tree/master/examples/pingserver reaches 3M qps on a single instance for redis-benchmark (ping_inline API) on c5n ec2 instances.

And where is your 1-to-1 comparison with epoll? Or you just go by feeling of "high numbers"? See my benchmark for a 1-to-1 comparison.

**Comment by alexhultman** [comment]
@axboe 

> So in short, I'm of course interested in cases where io_uring isn't performing to its full potential, it most certainly should. 

Woud you look at my new benchmark? I have eliminated everything but epoll and io_uring and on both my machines epoll wins despite io_uring being SQ-polled with 0 syscalls and using pre-registered file descriptors and buffers. I'm not involving any networking at all.

strace shows the epoll case make millions of syscalls while the io_uring is entirely silent in the syscalling department.

What am I doing wrong / why is io_uring not performing?

**Comment by alexhultman** [comment]
> @alexhultman Check netty/netty#10622 - I won't have any bandwidth for the next day or two, but maybe that will pique your interest.

The fact that a virtualized garbage collected JIT-stuttery Java project can see performance improvements when swapping from epoll to io_uring is not a viable proof of io_uring itself (as a kernel feature) being more performant than epoll itself (as a kernel feature). It really just proves that writing systems in non-systems programming languages are going to get you poor results.

io_uring does more things in kernel, meaning a swap from epoll to io_uring leads to less things happening in Java. As a general rule of tumb; the less you do in high level garbage collected virtualized code, the better.

**Comment by romange** [comment]
@alexhultman  which kernel version did you use?

**Comment by alexhultman** [comment]
It is clearly stated in the posted text. 5.9.9

**Comment by santigimeno** [comment]
@alexhultman By running your tests locally with a `5.10-rc5` version it seems I'm seeing `io_uring` behave better, or am I reading it wrong?:
```
$  ./epoll 1000
Pipes: 1000
Time: 16.059609

$ sudo ./io_uring 1000
Pipes: 1000
Time: 10.984051

$ ./epoll 1500
Pipes: 1500
Time: 24.501726

$ sudo ./io_uring 1500
Pipes: 1500
Time: 18.112729

$ ./epoll 2000
Pipes: 2000
Time: 37.705230

$ sudo ./io_uring 2000
Pipes: 2000
Time: 26.174995

```

**Comment by romange** [comment]
First of all, there is a bug that hurts io_uring performance.  its fix will be merged into 5.11. 

Secondly, I've looked at your benchmark code and you test something that is not necessarily relevant nor optimal for networking use-case.
1. You test `IORING_SETUP_SQPOLL` mode - I did not succeed to get any performance gain there with sockets. In fact it was consistently worse than using non-polling mode.

2. You test a single epoll/io_uring loop which does not trigger contention edge-cases inside kernel. When you have N cores running N epoll loops doing read/writes via socket you put 100% load on your machine, you will see how io_uring performs better.

Finally, I've never tried using pipes in my tests. io_uring essentially delegates requests to the their corresponding APIs. So if, for example, pipes kernel code takes most CPU you may indeed not see much difference by using io_uring or epoll.

**Comment by Qix-** [comment]
Also, just to add (note that I was a skeptic above as well): Hypervisors tend to level out epoll vs io_uring performance benchmarks in my findings - if you're within a VM, expect io_uring to have less than ideal performance increases against epoll. Testing on bare metal seems to make quite a bit of difference depending on how you're using io_uring.

> First of all, there is a bug that hurts io_uring performance. its fix will be merged into 5.11.

@romange Is that already on the io_uring branch of mainline? Just curious.

**Comment by martin-g** [comment]
Here are the results for @alexhultman 's benchmark on my machine (MyTuxedo laptop, Intel i7-7700HQ (8) @ 3.800GHz, 32GB RAM, Ubuntu 20.10 x86_64, Kernel: 5.10.0-051000rc6-generic)

```
 make runs
rm -f epoll_runs
rm -f io_uring_runs
for i in `seq 100 100 1000`; do ./io_uring $i; done
Pipes: 100
Time: 0.908056
Pipes: 200
Time: 2.063551
Pipes: 300
Time: 3.183146
Pipes: 400
Time: 4.810344
Pipes: 500
Time: 5.609743
Pipes: 600
Time: 8.197645
Pipes: 700
Time: 10.275732
Pipes: 800
Time: 11.889881
Pipes: 900
Time: 15.030963
Pipes: 1000
Time: 15.421023

for i in `seq 100 100 1000`; do ./epoll $i; done
Pipes: 100
Time: 1.575792
Pipes: 200
Time: 3.173769
Pipes: 300
Time: 5.173567
Pipes: 400
Time: 7.255583
Pipes: 500
Time: 10.283918
Pipes: 600
Time: 12.986523
Pipes: 700
Time: 14.560208
Pipes: 800
Time: 17.426127
Pipes: 900
Time: 19.796715
Pipes: 1000
Time: 23.262279
```

io_uring gives better results!

**Comment by romange** [comment]
> Also, just to add (note that I was a skeptic above as well): Hypervisors tend to level out epoll vs io_uring performance benchmarks in my findings - if you're within a VM, expect io_uring to have less than ideal performance increases against epoll. Testing on bare metal seems to make quite a bit of difference depending on how you're using io_uring.
> 
> > First of all, there is a bug that hurts io_uring performance. its fix will be merged into 5.11.
> 
> @romange Is that already on the io_uring branch of mainline? Just curious.

I do not think it's on io_uring branch because the fix was not there.
Here is the relevant article: https://www.phoronix.com/scan.php?page=news_item&px=Linux-5.11-Task-Work-Opt

**Comment by YoSTEALTH** [comment]
```bash
$ make
gcc -O3 epoll.c -o epoll
gcc -O3 io_uring.c /usr/lib/liburing.a -o io_uring
gcc: error: /usr/lib/liburing.a: No such file or directory
make: *** [Makefile:3: default] Error 1
```

**Comment by romange** [comment]
> ```shell
> $ make
> gcc -O3 epoll.c -o epoll
> gcc -O3 io_uring.c /usr/lib/liburing.a -o io_uring
> gcc: error: /usr/lib/liburing.a: No such file or directory
> make: *** [Makefile:3: default] Error 1
> ```

```
git submodule update --init --recursive 
cd liburing/
./configure && make
sudo make install
```

**Comment by alexhultman** [comment]
@martin-g @santigimeno That's very interesting - thanks for reporting! I will do some more testing on newer kernels and see if I can finally get to see this supposed io_uring wonder myself.

**Comment by alexhultman** [comment]
@romange You do a lot of confident talking but you still refuse to follow up with any actual testing of your claims.

1. I have tested all modes, the mode without SQ polling had insignificant differences in performance and because it caused syscalls to appear, I wanted to use SQ because that is where all the fuzz is about regarding io_uring.

2. You haven't tested this, but you just **assume**. I did an actual test of this and I got results significantly opposing your so confident assumption.

Please do testing before you make up assumptions about everything. This entire thread is about this exact behavior - show with actual numbers like @santigimeno and @martin-g did.

**Comment by axboe** [comment]
@alexhultman Please tone down the snark. FWIW, @romange has done plenty of testing in the past, and was instrumental in finding the task work related signal slowdown for threads. Questioning the validity of a test is _perfectly_ valid, in fact it's the very first thing that should be done before potentially wasting any time on examining results from said test. Nobody is in a position to be making any demands in here, in my experience you get a lot further by being welcoming and courteous.

**Comment by alexhultman** [comment]
> When you have N cores running N epoll loops doing read/writes via socket you put 100% load on your machine, you **will** see how io_uring performs better.

The above is not a question, it is a confident statement without any backing proof other than "it will be the case". This is what is the issue here - blindly making claims without any backing proof other than "listen to my assumption".

**Comment by alexhultman** [comment]
> Nobody is in a position to be making any demands in here

Nobody is demanding anything here.

**Comment by axboe** [comment]
> @romange Is that already on the io_uring branch of mainline? Just curious.

The task work fixes are a bit split - the core and x86 of them are queued up in the x86 tree, I did the rest of the archs and the cleanups on top in my ```tif-task_work.arch``` branch here:

https://git.kernel.dk/cgit/linux-block/log/?h=tif-task_work.arch

It's based on the x86/core/generic tree. All of it will go into 5.11 once the merge window for that opens.

**Comment by alexhultman** [comment]
Did anyone test with mitigations off, or a modern CPU with hardware mitigations? Syscalling overhead is way lower with mitigations off.

**Comment by axboe** [comment]
I took a quick look, I'll check do more in-depth once I have more time. Haven't tested any of this, just thoughts as I checked what was being done. What I found:

- The registered buffers may seem like a good idea, but for pipes it'd just mean that you're hitting a presumably slower memcpy rather than the faster copy_to_user. Registered buffers are only an optimization if you're actually doing DMA on them, which this test case is not (I'd be surprised if you aren't just copy bound). Would make more sense to remove the registered buffers, and use io_uring_prep_read() and io_uring_prep_write() for the buffers.

- Buffers are 1k, I'd use something smaller if you're interested in benchmarking the interface rather than just a copy. 32b or something would be more suitable.

- You set SQPOLL and SQ_AFF, but you don't actually set the affinity of the thread. Hence it's zero and you're binding to CPU0 every time. May not matter, but it's somewhat nonsensical to ask for binding and then not set a preferred target. Since with SQPOLL you're coordinating with the kernel thread, for best results you'd want to ensure that you're user task and kernel thread have a sane setup in terms of the machine/CPU topology.

**Comment by alexhultman** [comment]
Thanks for hints. Although I have tested all of this before - initially I did 6 byte messages and no fixed buffers and did not use SQ polling or affinity.

The only reason I added all of that was because I was not getting the results I wanted. I did get sightly better perf. without them but still lower than epoll either way.

**Comment by axboe** [comment]
FWIW, ran it on my AMD box. It's a 3970x, zen2 based. I have no mitigations enabled:

```
root@amd:/home/axboe/git/io_uring_epoll_benchmark# taskset -c 0 ./io_uring 100
Pipes: 100
Time: 0.299408
root@amd:/home/axboe/git/io_uring_epoll_benchmark# taskset -c 0 ./io_uring 4000
Pipes: 4000
Time: 17.003238
```

```
root@amd:/home/axboe/git/io_uring_epoll_benchmark# taskset -c 0 ./epoll 100
Pipes: 100
Time: 0.336690
root@amd:/home/axboe/git/io_uring_epoll_benchmark# taskset -c 0 ./epoll 4000
Pipes: 4000
Time: 20.755568
```

This is using 5.10-rc5 (ish).

**Comment by alexhultman** [comment]
Thanks a lot for the measurements!

**Comment by Qix-** [comment]
Alternatively to get build of the benchmark working without a global install, patch makefile:

```patch
diff --git a/Makefile b/Makefile
index 537d8c4..03edd8c 100644
--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,8 @@
 default:
+       cd liburing && ./configure
+       make -C liburing
        gcc -O3 epoll.c -o epoll
-       gcc -O3 io_uring.c /usr/lib/liburing.a -o io_uring
+       gcc -O3 io_uring.c ./liburing/src/liburing.a -I./liburing/src/include -o io_uring

 runs:
        rm -f epoll_runs
```

**Comment by alexhultman** [comment]
I've compiled 5.10rc7 and rerun tests - I still get favorable results for epoll. However if I enable all mitigations, epoll takes a massive hit while io_uring takes much less of a hit. Now io_uring is faster (by about 2x), but this really is just epoll syscalls being completely crippled by software workarounds to old Intel bugs that are fixed in hardware for new CPUs. IO_uring should outperform epoll even without mitigations for it to be an architectural win and not just because Intel CPUs are riddled with security issues.

Remember that I did test with mitigations long ago (see initial post) so at least I can now see io_uring outrunning epoll when having mitigations enabled - but still not without them. Syscalls without the massive burden of mitigations are still faster for me.

**Comment by romange** [comment]
I've created a ping-pong loadtesting tool that checks both epoll and io_uring. The code is almost the same in both setups besides the event-loops and networking code that calls them.

The main tool is here: https://github.com/romange/async/blob/master/examples/echo_server.cc
To run in server mode:
`./echo_server  --backlog 2048` for iouring and `./echo_server  --backlog 2048  --epoll` for epoll.


I've run the benchmark on AWS, c5n.9xlarge with 36 virtual CPUs hosting a server process.

I used 2 other c5n.9xlarge machines to send traffic:
./echo_server --connect 10.0.0.1   -c 100 -n 100000  --proactor_threads=32 --size=1` which spans 32 io threads with 100 tcp connection in each.

epoll was consistently slower by 10% (see the attached screenshots)
![Screenshot from 2020-12-09 15-23-42](https://user-images.githubusercontent.com/3674760/101636724-6741ad80-3a34-11eb-8431-55d5f24ba392.png)
![Screenshot from 2020-12-09 15-21-21](https://user-images.githubusercontent.com/3674760/101636743-6dd02500-3a34-11eb-8f57-0d47e03dcd1c.png)
![Screenshot from 2020-12-09 15-23-16](https://user-images.githubusercontent.com/3674760/101636764-745e9c80-3a34-11eb-9983-e14794c85fa7.png)

I admit that 10% is not a deal-breaker for pure network workloads. It probably makes sense to use io_uring for use-cases where one requires an efficient networking and disk I/O.

Both setups were bottlenecked on CPU, more exactly on system cpu as you can see in top output.
Reaching

**Comment by axboe** [comment]
I changed the size count to 32b as described, and ran it without registered buffers (also as described). Here's the raw data:

```

Pipes  | io_uring  | epoll
-----------------------------
10     | 0.031589  | 0.030885
50     | 0.146127  | 0.148299
100    | 0.290833  | 0.298824
250    | 0.732903  | 0.761741
500    | 1.47817   | 1.564801
1000   | 3.012716  | 3.158615
2000   | 6.465537  | 7.510465
4000   | 17.334865 | 20.301635
8000   | 38.496499 | 45.603121
```

And graphed:

![io_uring and epoll (time spent, lower is better)](https://user-images.githubusercontent.com/2619634/101666681-be7f5800-3a0b-11eb-8ca0-811296d7ab13.png)

Test system is the same as earlier, 3970X, and all tests are pinned to CPU0 to reduce variability. It's running 5.10-rc7 with whatever I have pending for 5.11 merged in, though I don't expect there to be differences between 5.10-rc7. There is NO mitigations enabled, (retpoline is off, page table isolation is off, etc). For a belt and suspenders kind of thing, it's also booted with ```mitigations=off``` added to the kernel command line.

**Comment by axboe** [comment]
> I admit that 10% is not a deal-breaker for pure network workloads

People get bent out of shape for way smaller wins than that, so I'd say 10% is a pretty big deal!

**Comment by Qix-** [comment]
Right, it's more just the initial claims that don't match the common case that I think warranted the response.

A 10% increase is well beyond what I'd personally consider "worth it", especially if it's within the realm of 'common case' and not an edge case (like some of the higher figures before seem to be).

The one thing I'm a bit confused about is the claim made above that SQPOLL doesn't add any performance - I'd like to know under which circumstances that was the case, as it requires a specific setup to use efficiently.

**Comment by alexhultman** [comment]
@romange your code is full of abstractions, boost and other known to be bloated libraries that direct use of epoll will outperform. If you aim to benchmark a kernel feature you should minimize to only syscalls directly with no wrapper between. I know this intimately because my use of epoll is much faster than boost and libuv and such poorly written wrappers

**Comment by alexhultman** [comment]
A kernel feature benchmark should be of nothing but the actual syscalls (or well, glibc). Using abstractions like complete libraries is not accurate and referring to Netty and Node.js is so wrong it's hard to even begin to point out how incredible inaccurate such examples are

**Comment by romange** [comment]
My code has abstractions, yes and I never claimed it's a pure load-test that is benchmarking a single syscall.
I was never interested in benchmarking a simple system call.

Having said that, the event-loop in `echo_server` is implemented using my own code without boost-asio or other third-party libraries. I needed relatively complex logic in order to code a server that spans multiple cores and can reach hardware limits with respect to networking and CPU. 

Your original question was
```
And where is your 1-to-1 comparison with epoll? Or you just go by feeling of "high numbers"? See my benchmark for a 1-to-1 comparison.
```

I think this program gives you the answer to whether my claims based on a "feeling"  or on hard data and whether I am "blindly making claims" :) 

If you feel my code might be inefficient, bloated and thus skews results**, (which is a reasonable concern  )I suggest that you run it first on your computer and compare to your implementation. I invested lots of effort to write an efficient networking engine and I hope it's close to the practical limits of what can be achieved via kernel networking stack.  

** My only boost dependency is Boost.Fibers - which are very well written and have pretty good performance. Their CPU usage  is relatively low and heavily dominated by epoll/io_uring CPU therefore I do not think they affect the final result.

**Comment by romange** [comment]
> > I admit that 10% is not a deal-breaker for pure network workloads
> 
> People get bent out of shape for way smaller wins than that, so I'd say 10% is a pretty big deal!

I did not mean to diminish io_uring accomplishments. Moreover, my benchmark uses exactly one read()+write() per request which is often not the case for real production cases. For patterns that use more syscalls the gap would probably wider.  

I was referring mostly to the existing software that has lots of dependency on epoll(). Betweem rewriting low-level layers in such codebases and reserving 10% more hardware, I would choose increasing hardware allocations. For companies like Facebook 10% is significant savings but for small startups the rewrite investment will probably more expensive.

**Comment by twissel** [comment]
And let's not forget the initial design goal for io_uring:  "fast async file io".

**Comment by cryptoquick** [comment]
I have to say, this has been a fascinating read and I'm eagerly anticipating the conclusion. I feel like there may be something of importance to be learned here, but I'm not yet a good enough technologist to know what that is.

I do appreciate the spirit of the debate, and encourage it, but civility is important for some in order to get things accomplished. I know I burn a lot more cycles if I'm frustrated that could then be spent towards problem-solving. I wonder if the tone had been a bit less confrontational, this might be done by now. But I worry that if it hadn't, this may never have gotten the attention it should. There's something to be learned from what's happening here, and I wish to learn it.

In the future, I think places like these are where important things are realized by the broader community of technologists and can be understood and appreciated by many. A more rigorous approach, with a measure of humility and maturity is also warranted, without a doubt. Just know, there may eventually be more eyes on what takes place here at some point in the future, provided that there is merit in what is found, and if so... Well, I suppose I'd recommend conducting any participation in this discussion as if that is indeed the case.

For what it's worth, I arrived at this thread because it ranked highly in search results for "epoll vs io_uring", in an effort to better understand the difference.

**Comment by morganthomas** [comment]
I replicated @alexhultman 's result, on Linux 5.10.2, NixOS 20.09, on an HP Envy TE01-0014 with an Intel Core i5-8400 and 32 GB RAM and an HP brand SSD.

With hardware mitigations enabled, io_uring is faster:

```
[morgan@ripley:~/code/io_uring_epoll_benchmark]$ sudo ./io_uring 100
Pipes: 100
Time: 0.600844

[morgan@ripley:~/code/io_uring_epoll_benchmark]$ sudo ./epoll 100
Pipes: 100
Time: 1.289989
```

Using the boot parameter `mitigations=off` to disable all hardware security vulnerability mitigations, epoll is faster in this benchmark (and it segfaults on exit to boot):

```
[morgan@ripley:~/code/io_uring_epoll_benchmark]$ ./epoll 100
Pipes: 100
Time: 0.510842
Segmentation fault (core dumped)

[morgan@ripley:~/code/io_uring_epoll_benchmark]$ sudo ./io_uring 100
[sudo] password for morgan: 
Pipes: 100
Time: 0.547766
```

**Comment by morganthomas** [comment]
For what it's worth, I was able to get rid of the segfault on the epoll benchmark (under the same test conditions) by running the benchmark as root, and epoll is still faster:

```
[morgan@ripley:~/code/io_uring_epoll_benchmark]$ sudo ./io_uring 100
Pipes: 100
Time: 0.553995

[morgan@ripley:~/code/io_uring_epoll_benchmark]$ sudo ./epoll 100
Pipes: 100
Time: 0.524630
```

**Comment by NobodyXu** [comment]
@alexhultman There's a serious problem with your code at [alexhultman/io_uring_epoll_benchmark](https://github.com/alexhultman/io_uring_epoll_benchmark):

```
/* Submit */
for (int j = 0; j < num_pipes * 2; j += io_uring_submit(&ring));
```

This is not the right way to utilize `io_uring`. `io_uring` gives you the ability to submit patches of IO request in one syscall:

>  int io_uring_submit(struct io_uring *ring)
> 
>     Submits the SQEs acquired via io_uring_get_sqe() to the kernel. You can call this once after you have called 
> io_uring_get_sqe() multiple times to set up multiple I/O requests.
> 
>     Parameters
> 
>         ring: io_uring structure as set up by io_uring_queue_init().
> 
>     Return value: returns the number of SQEs submitted.

**Comment by alexhultman** [comment]
> This is not the right way to utilize io_uring. io_uring gives you the ability to submit patches of IO request in one syscall

You haven't understood the code you highlighted. It **does** only submit once. It only submits multiple times if it **failed** to submit everything in one go.

The rest, we have already covered in the thread of you would like to have a read.

We basically concluded so far that io_uring can outperform epoll when mitigations are on (it does for me), but with mitigations off epoll wins (that's the basis for the graph).

> No wonder io_uring's performance for you have suck as hell.

Ah yes I'm such a moron, aren't I?

**Comment by NobodyXu** [comment]
@alexhultman Also:

```
/* Did we really read/write everything? */
if (cqes[j]->res != BUF_SIZE) {
    printf("Mismatching read/write: %d\n", cqes[j]->res);
    return 0;
}
```

> It is not an error if this number is smaller than the number of bytes requested; this may happen for example because fewer bytes are actually available right now (maybe because we were close to end-of-file,
> or because we are reading from a pipe, or from a terminal), or because read() was interrupted by a signal.  See also NOTES.

The above reference is taken from `man 2 read`.

If syscall `read` or `write` returns less than requested, it's not an error. So same for `read` or `write` from `io_uring`.

What you are doing here is totally wrong.

**Comment by alexhultman** [comment]
Again, you fail to understand the code you're referring to.

Please, by all means, send me a PR to the benchmark and I will gladly run it and give you a personal back massage by my very hands if you can correct it so that my machine does in fact outperform epoll.

Please, do it. I'm open. Go ahead.

**Comment by NobodyXu** [comment]
> You haven't understood the code you highlighted. It does only submit once. It only submits multiple times if it failed to submit everything in one go.
> 
> The rest, we have already covered in the thread of you would like to have a read.
> 
> We basically concluded so far that io_uring can outperform epoll when mitigations are on (it does for me), but with mitigations off epoll wins (that's the basis for the graph).

About that, I seemed to be wrong.

But still, I wonder why are you calling `io_uring_submit` that many times at all.

Given that once the kernel polling thread is started, there is no need to `io_uring_submit` and the userspace should be able to read from CQE directly unless it wants to wait for `min_complete.

**Comment by NobodyXu** [comment]
> Again, you fail to understand the code you're referring to.
> 
> Please, by all means, send me a PR to the benchmark and I will gladly run it and give you a personal back massage by my very hands if you can correct it so that my machine does in fact outperform epoll.
> 
> Please, do it. I'm open. Go ahead.

What I am saying is that, `cqes[j]->res != BUF_SIZE` is not an error.

**Man page of `read` explicitly say this is not an error**, so treating it as an error here is wrong.

**Comment by NobodyXu** [comment]
> > But still, I wonder why are you calling io_uring_submit that many times at all.
> 
> io_uring_submit returns the number of entries submitted. So because I know exactly how many I have allocated and filled out, I check so that I actually do send them all off. This to make sure the benchmark does what it is supposed to do and not just short circuit.

You can check that by counting CQE entries, way faster than try to use `io_uring_submit` for that.

The whole point of SQ polling mode is to avoid syscall and use CQE to retrieve information in the userspace.

**Comment by alexhultman** [comment]
> What I am saying is that, cqes[j]->res != BUF_SIZE is not an error.

> Man page of read explicitly say this is not an error, so treating it as an error here is wrong.

What man pages say is an "error" is entirely out of context in regards to what is acceptable for this benchmark.

Because I send a fixed set of bytes I need to make sure I get said set of bytes. This benchmark assumes (and measures) the **fast path** of epoll vs the fast path of io_uring so no user space buffering or retries can be part of this run.

That's why the check is

**Comment by NobodyXu** [comment]
Note that the pipe has a limit. If you send more than that limit, it might only write what it could.

Also, note that `io_uring` submits io **in any order it prefers** unless linking is used.

So it is completely possible for your read requests to all happen before write requests.

**Comment by alexhultman** [comment]
> You can check that by counting CQE entries, way faster than try to use io_uring_submit for that.

So make a PR and prove your point scientifically instead of just **claiming** this or that is "way faster".

> The whole point of SQ polling mode is to avoid syscall and use CQE to retrieve information in the userspace.

You obviously haven't even tested to run under strace. Please do your homework before claiming a bunch of sensational nonsense

**Comment by NobodyXu** [comment]
IMHO the solution I can think of is to link together one read and one write as one link.

This would ensure the pipe is never overflowed.

**Comment by NobodyXu** [comment]
> > You can check that by counting CQE entries, way faster than try to use io_uring_submit for that.
> 
> So make a PR and prove your point scientifically instead of just **claiming** this or that is "way faster".
> 
> > The whole point of SQ polling mode is to avoid syscall and use CQE to retrieve information in the userspace.
> 
> You obviously haven't even tested to run under strace. Please do your homework before claiming a bunch of sensational nonsense

I am not saying this or that is faster.

_**I am saying your code is invalid unix code, so the benchmark couldn't count**_

**Comment by alexhultman** [comment]
> I am not saying this or that is faster.

And

> You can check that by counting CQE entries, way faster than try to use io_uring_submit for that.

Are two logically contradictory statements

**Comment by NobodyXu** [comment]
> > I am not saying this or that is faster.
> 
> And
> 
> > You can check that by counting CQE entries, way faster than try to use io_uring_submit for that.
> 
> Are two logically contradictory statements


```
/* Poll for completions */
int remaining_completions = num_pipes * 2;
while (remaining_completions) {

    /* Poll until done in batches */
    int completions = io_uring_peek_batch_cqe(&ring, cqes, MAX_BATCH);
    if (completions == 0) {
        continue;
    }

    /* Check and mark every completion */
    for (int j = 0; j < completions; j++) {

        /* Check for failire */
        if (cqes[j]->res < 0) {
            printf("Async task failed: %s\n", strerror(-cqes[j]->res));
            return 0;
        }

        /* Did we really read/write everything? */
        if (cqes[j]->res != BUF_SIZE) {
            printf("Mismatching read/write: %d\n", cqes[j]->res);
            return 0;
        }

        /* Mark this completion as seen */
        io_uring_cqe_seen(&ring, cqes[j]);
    }

    /* Update */
    remaining_completions -= completions;
}
```

You already has the code for counting CQE entries, so it would definitely be faster if you avoid using

```
for (int j = 0; j < num_pipes * 2; j += io_uring_submit(&ring));
```

at all.

**Comment by NobodyXu** [comment]
Besides, it's not the usually userspace's task to check whether the kernel has be lazy.

No sane application do that.

If you think the `io_uring` code is faking but skipping some entries, you should make a separate executable for checking that.

Having this code in your performance critical path definitely influence your performance to some extent.

> @romange your code is full of abstractions, boost and other known to be bloated libraries that direct use of epoll will outperform. If you aim to benchmark a kernel feature you should minimize to only syscalls directly with no wrapper between. I know this intimately because my use of epoll is much faster than boost and libuv and such poorly written wrappers

If you do not like using abstractions, then why not also remove this since this checking also influences your performance?

**Comment by alexhultman** [comment]
I'm a scientist. Please stop guessing/claiming, and just give me a PR already. Then we can see how fantastically superior your use of io_uring is. I'm honestly meaning it - if you can produce something that changes my mind I'm going to be very thankful.

Until then, I'm going to continue writing "invalid Unix code".

**Comment by alexhultman** [comment]
> IMHO the solution I can think of is to link together one read and one write as one link.
> 
> This would ensure the pipe is never overflowed.

Holy shit dude. Have you even looked at the code!? That's exactly what it does.....

**Comment by alexhultman** [comment]
> you should make a separate executable for checking that.

Ah yes I'll just make a separate process and do IPC between them for checking an integral value. Wow, fantastic advice!! Swapping a branch predicted conditional jump based on a register value for a full on IPC between two separately scheduled processes.

What more, sensei?

**Comment by alexhultman** [comment]
> Besides, it's not the usually userspace's task to check whether the kernel has be lazy.

> No sane application do that.

Literally all applications do that. It's basic backpressure management. You always have to make sure the kernel has received what you give it, before you toss it. **At least when only looking at the fast path**. Which I'm doing.

If you care for the **slow path** you also have to take into account backpressure buffering (which io_uring should be faster at).

But this benchmark only cares to test for the **fast path** and here epoll wins.

**Comment by alexhultman** [comment]
@NobodyXu 

Dearest Mr. Nobody.

Because I am a scientist who **tests ideas with reality as my judge** and not just **makes nonsensical shit up with no testing, whatsoever**, I decided to follow your so all-knowing expertise on this.

> You already has the code for counting CQE entries, so it **would definitely be faster** if you avoid using

> for (int j = 0; j < num_pipes * 2; j += io_uring_submit(&ring));

> at all.

----

> There's a serious problem with your code at alexhultman/io_uring_epoll_benchmark:

> /* Submit */
> for (int j = 0; j < num_pipes * 2; j += io_uring_submit(&ring));

> This is not the right way to utilize io_uring. [...]

----

> You can check that by counting CQE entries, **way faster than try to use io_uring_submit for that**.


----

Your supposed "way faster" change **entirely broke the test, with no progress made whatsoever**. Let it be clear that you have not the slightest idea about what you're talking. You haven't even tested your idea, you haven't even tested my idea, you have tested nothing at all. And you just say a bunch of shit with confidence. You are the complete opposite of a scientific approach and your input is entirely useless (as proven by testing).

Thank you

**Comment by isilence** [comment]
@NobodyXu, that "for" loop is fine, should call submit() once in general.
The "BUF_SIZE" is also fine as long as the test doesn't fail on it.

@alexhultman, frankly, that's getting bloody annoying.

> We basically concluded so far that io_uring can outperform epoll when mitigations are on (it does for me), but with mitigations off epoll wins (that's the basis for the graph).

Seems it's only you who concluded it. There are mixed results above for your own benchmark with mitigations=off, see above. It would be also interesting to see numbers for other media, SSD, block devices, so on, not just pipes. May yield a similar trend, but at least it would back your conclusions, which at the moment are too broad comparing to the testing you've done.

Regarding the benchmark, there are places I don't like. May be better to fix/experiment on them.
1. You use a single buffer for all reads and writes. Minor and depends much, but better to allocate e.g. per pair.
2. You don't keep a fixed QD but do them in waves, that's a strange pattern
3. io_uring_peek_batch_cqe() does unnecessary copies, I'd replace it with io_uring_for_each_cqe()
4. CQ polling with io_uring_peek_batch_cqe() can potentially interfere with SQPOLL task taking its cycles, try out io_uring_wait_cqes() (+io_uring_for_each_cqe()) to see if there is any difference. Or to spread affinity.
5. Did you try to vary BUF_SIZE? SQPOLL will copy all BUF_SIZE * N bytes in one thread, but may be that the epoll do it less efiiciently but in parallel (internally in the kernel). Didn't read through the epoll benchmark, but would be interesting to see CPU load or comparison to IOSQE_ASYNC.

For not SQPOLL (if you test it), there is also a chance that all reqs just complete inline (i.e. during submission). Would be curious to see how many CQEs the first io_uring_peek_batch_cqe() call returns. This can affect performance greatly, and e.g. in that case swapping write/read submission order should pessimise it.

**Comment by alexhultman** [comment]
How annoying do you think it is for me? When people just come in and say a bunch of unconfirmed nonsense that doesn't even add up.

But as always, I'm very much interested in anything that can improve io_uring results in this test. But it has to be backed by working code, can't just be random emotions and overly strong confidence that don't even run. I always look for things to change my mind, rather than look for things to confirm what I already have decided to believe in.

**Comment by isilence** [comment]
> But as always, I'm very much interested in anything that can improve io_uring results in this test. But it has to be backed by working code, can't just be random emotions and overly strong confidence that don't even run. I always look for things to change my mind, rather than look for things to confirm what I already have decided to believe in.

Let's agree on that and put down spamming. As I wrote before, I'm interested myself to see when we stand for all different use cases.

BTW, (5) is probably is not the case, so don't mind it

**Comment by Qix-** [comment]
Folks, conversation discussing the third party benchmark should be taken to the respective repository and xref'd from here if anything. Let's refrain from further discussing it on this issue, as it generates considerable email noise for those watching.

This is a public academic debate, after all. Let's be respectful to those watching.

**Comment by NobodyXu** [comment]
@isilence `io_uring_for_each_cqe()` doesn't move `*(ring)->cq.khead` forward and might risk overflowing CQE:

```
#define io_uring_for_each_cqe(ring, head, cqe)              \           
    /*                              \                                   
     * io_uring_smp_load_acquire() enforces the order of tail   \       
     * and CQE reads.                       \                           
     */                             \                                   
    for (head = *(ring)->cq.khead;                  \                   
         (cqe = (head != io_uring_smp_load_acquire((ring)->cq.ktail) ? \
        &(ring)->cq.cqes[head & (*(ring)->cq.kring_mask)] : NULL)); \   
         head++)                            \                     
```

Perhaps it is better to use `io_uring_peek_cqe` and `io_uring_cqe_seen` to avoid copies of CQE?

**Comment by isilence** [comment]
@NobodyXu, `io_uring_for_each_cqe()` is the fastest (well, almost...) to do the job. Go look for examples, e.g. under `test/` or for `io_uring_cq_advance()`.

**Comment by NobodyXu** [comment]
@isilence Thanks for the advice, I will take a look.

**Comment by morganthomas** [comment]
Could anybody give me a pointer on running the first party benchmark? As far as I know, I am the only person who has tried to reproduce Alex Hultman's result under the conditions he specified, with hardware mitigations off, and I was able to do so. Now the controversy is about the benchmark he wrote; but he said he obtained the same result with the first party benchmark with mitigations off. If nobody else is going to try to replicate his results and follow his instructions to do so, then I suppose I might as well.

**Comment by alexhultman** [comment]
Just boot with mitigations=off and run sudo ./io_uring 100 or any number. Do the same for .epoll

**Comment by morganthomas** [comment]
Alex, I ran your benchmark. I'm wondering how do I run @axboe 's benchmark? I didn't find anything on that when I did a little bit of research in December.

**Comment by alexhultman** [comment]
I don't think he has a benchmark for this case? There are third party benchmarks for sockets but on both my machines epoll wins there as well, for all three I have found.

**Comment by isilence** [comment]
@morganthomas, I believe Jens run that same benchmark. If some other were
mentioned, it's probably fio `fio/t/io_uring` or fio itself.

BTW, running for 0.5s doesn't sound right, did you average across multiple runs
or something? The difference is not large enough to not blame e.g. your CPU
governor with turbo boost kicking in. Just want check that the method is right.

@alexhultman, if benchmark do what I think, I can cut ~30-40% of io_uring
kernel overhead, partially patches were posted before. What, I'm more curious
about is whether there are real apps behaving like that, or it's benchmarking
just for benchmarking.

**Comment by alexhultman** [comment]
> if benchmark do what I think, I can cut ~30-40% of io_uring
kernel overhead, partially patches were posted before.

Cool. Please share your variant and I'll run it

**Comment by isilence** [comment]
> > if benchmark do what I think, I can cut ~30-40% of io_uring
> > kernel overhead, partially patches were posted before.
> 
> Cool. Please share your variant and I'll run it

To be clear:
1. it's about the kernel side of things, for the exactly same benchmark.
2. It's io_uring overhead, so absolute numbers would depend on the share
of time it sits in pipe_[read,write].

**Comment by alexhultman** [comment]
So are you looking to submit this change to the kernel? I would love that change.

> What, I'm more curious
about is whether there are real apps behaving like that, or it's benchmarking
just for benchmarking.

It's not benchmarking just for benchmarking. I run a small business doing IO for a handful multi-billion dollar companies relying on small message sending performance. So this change would mean I could finally see benefits from using io_uring which would be fantastic.

I believe I have "isolated" the kind of performance I would like to see improved in io_uring, via this simple benchmark (even if the benchmark itself is ridiculous).

**Comment by alexhultman** [comment]
@isilence Btw, if you have a patch please send me it and I can compile this modified kernel and test it out

**Comment by isilence** [comment]
@alexhultman, I was going for it anyway (maybe for-5.13), the benchmark
just might fall under conditions, will see.

I'll give it a run, but that's for when I allote time to cook it right.

**Comment by alexhultman** [comment]
Perfect! I feel like we are making progress on this issue now!

**Comment by axboe** [comment]
I ran the tests on the current for-5.12/io_uring branch. For full disclosure, I made the following changes to the io_uring test:

- Don't register the buffers. Won't make a runtime difference for the actual loop for this test case, just removes a tiny bit of startup time. Since the registered buffers don't make a difference for the test (reading/wring to/from pipes), it doesn't make sense to do it.
- Use the non-fixed versions of reading/writing, which is necessary due to not using registered buffers.
- Using one buffer like the epoll case. This is more apples-to-apples, as otherwise tons of pipes have much worse memory locality which will skew the results.
- Remove the io_uring_submit() loop. We don't need to loop, just submit it.

That said, I ran with 4000 pipes, using a BUF_SIZE of 32b, 1k, and 4k. Just to get an idea of what the different sizes would look like. The kernel used has NO mitigations enabled (PAGE_TABLE_ISOLATION off, RETPOLINE off, booted with mitigations=off). This should be favorable to epoll, as it does more syscalls. Raw results:

```
# time taskset -c 0 ./epoll 4000
Pipes: 4000, BUF_SIZE=32
Time: 22.889015

real	0m22.941s
user	0m1.650s
sys	0m21.253s

# time taskset -c 0 ./epoll 4000
Pipes: 4000, BUF_SIZE=1024
Time: 34.387296

real	0m34.469s
user	0m1.630s
sys	0m32.772s

# time taskset -c 0 ./epoll 4000
Pipes: 4000, BUF_SIZE=4096
Time: 44.798268

real	0m44.917s
user	0m1.722s
sys	0m43.088s

# time taskset -c 0 ./io_uring 4000
Pipes: 4000, BUF_SIZE=32
Time: 16.104513

real	0m16.133s
user	0m0.576s
sys	0m15.536s

# time taskset -c 0 ./io_uring 4000
Pipes: 4000, BUF_SIZE=1024
Time: 25.663238

real	0m25.708s
user	0m0.633s
sys	0m25.040s

# time taskset -c 0 ./io_uring 4000
Pipes: 4000, BUF_SIZE=4096
Time: 32.218501

real	0m32.271s
user	0m0.645s
sys	0m31.583s
```

**Comment by axboe** [comment]
Same test, but this time enable mitigations and boot with mitigations=on. Only ran the 32b test case:

```
# time taskset -c 0 ./epoll 4000
Pipes: 4000, BUF_SIZE=32
Time: 27.012196

real	0m27.103s
user	0m1.668s
sys	0m25.360s

# time taskset -c 0 ./io_uring 4000
Pipes: 4000, BUF_SIZE=32
Time: 17.712587

real	0m17.757s
user	0m17.713s
sys	0m0.007s
```

io_uring ran 10% slower with io_uring off, epoll ran 18% slower.

**Comment by axboe** [comment]
Pavel suggested running 5.11 numbers as well, so here are the run from two comments up using 5.11 instead. Mitigations off again, to compare to those three 32/1k/4k runs:

```
# time taskset -c 0 ./io_uring 4000
Pipes: 4000, BUF_SIZE=32
Time: 17.249195

real	0m17.454s
user	0m17.250s
sys	0m0.007s

# time taskset -c 0 ./io_uring 4000
Pipes: 4000, BUF_SIZE=1024
Time: 27.043406

real	0m27.079s
user	0m27.044s
sys	0m0.007s

# time taskset -c 0 ./io_uring 4000
Pipes: 4000, BUF_SIZE=4096
Time: 35.807203

real	0m35.839s
user	0m35.808s
sys	0m0.006s
```

Still handily faster than epoll, but nice to see the improvements queued for 5.12 making a noticeable difference here.

**Comment by alexhultman** [comment]
Great, do you want to send me a PR for your best case and I'll run it on my machines

**Comment by axboe** [comment]
The diff is below. The SQPOLL comment in/out was for experimentation too, I'm not seeing much of a change in perf with that. Which is to be expected in this particular case.

```
diff --git a/epoll.c b/epoll.c
index 8da5449ad1d9..55fa6438faa5 100644
--- a/epoll.c
+++ b/epoll.c
@@ -13,7 +13,7 @@
 #include <string.h>
 
 #define MAX_PIPES 10000
-#define BUF_SIZE 1024
+#define BUF_SIZE 32
 
 int *pipes;
 
@@ -53,7 +53,7 @@ int main(int argc, char **argv) {
 
     int num_pipes = 0;
     sscanf(argv[1], "%d", &num_pipes);
-    printf("Pipes: %d\n", num_pipes);
+    printf("Pipes: %d, BUF_SIZE=%d\n", num_pipes, BUF_SIZE);
 
     int epfd = epoll_create1(0);
     for (int i = 0; i < num_pipes; i++) {
diff --git a/io_uring.c b/io_uring.c
index 2bdc5ed31180..6fe359c24902 100644
--- a/io_uring.c
+++ b/io_uring.c
@@ -16,7 +16,7 @@
 
 #define MAX_BATCH 2000
 
-#define BUF_SIZE 1024
+#define BUF_SIZE 32
 struct iovec buffer;
 
 #define MAX_PIPES 10000
@@ -30,7 +30,7 @@ int main(int argc, char **argv) {
 
     /* Pipes is first arg */
     sscanf(argv[1], "%d", &num_pipes);
-    printf("Pipes: %d\n", num_pipes);
+    printf("Pipes: %d, BUF_SIZE=%d\n", num_pipes, BUF_SIZE);
     for (int i = 0; i < num_pipes; i++) {
         if (pipe2(&pipes[i * 2], O_NONBLOCK | O_CLOEXEC)) {
             printf("Cannot create pipes!\n");
@@ -40,7 +40,10 @@ int main(int argc, char **argv) {
 
     /* Create an io_uring */
     struct io_uring ring;
-    if (io_uring_queue_init(5000, &ring, IORING_SETUP_SQPOLL | IORING_SETUP_SQ_AFF)) {
+    struct io_uring_params p = { };
+    //p.flags = IORING_SETUP_SQPOLL | IORING_SETUP_SQ_AFF;
+    //p.sq_thread_cpu = 1;
+    if (io_uring_queue_init_params(16384, &ring, &p)) {
         printf("Cannot create an io_uring!\n");
         return 0;
     }
@@ -52,14 +55,9 @@ int main(int argc, char **argv) {
     }
 
     /* Create and register buffer with io_uring */
-    buffer.iov_base = malloc(BUF_SIZE * 2 * num_pipes);
-    buffer.iov_len = BUF_SIZE * 2 * num_pipes;
-    memset(buffer.iov_base, 'R', BUF_SIZE * 2 * num_pipes);
-
-    if (io_uring_register_buffers(&ring, &buffer, 1)) {
-        printf("Cannot register buffer with io_uring!\n");
-        return 0;
-    }
+    buffer.iov_base = malloc(BUF_SIZE);
+    buffer.iov_len = BUF_SIZE;
+    memset(buffer.iov_base, 'R', BUF_SIZE);
 
     struct io_uring_sqe *sqe;
     struct io_uring_cqe **cqes = malloc(sizeof(struct io_uring_cqe *) * MAX_BATCH);
@@ -77,21 +75,21 @@ int main(int argc, char **argv) {
                 return 0;
             }
 
-            io_uring_prep_write_fixed(sqe, j * 2 + 1, buffer.iov_base + BUF_SIZE * (j * 2 + 1), BUF_SIZE, 0, 0);
+            io_uring_prep_write(sqe, j * 2 + 1, buffer.iov_base, BUF_SIZE, 0);
             sqe->flags |= IOSQE_FIXED_FILE;
 
-            /* Prepare a read */
+            /* Prepare a write */
             sqe = io_uring_get_sqe(&ring);
             if (!sqe) {
                 printf("Cannot allocate sqe!\n");
                 return 0;
-            }
-            io_uring_prep_read_fixed(sqe, j * 2, buffer.iov_base + BUF_SIZE * j * 2, BUF_SIZE, 0, 0);
+	    }
+
+            io_uring_prep_read(sqe, j * 2, buffer.iov_base, BUF_SIZE, 0);
             sqe->flags |= IOSQE_FIXED_FILE;
         }
 
-        /* Submit */
-        for (int j = 0; j < num_pipes * 2; j += io_uring_submit(&ring));
+	io_uring_submit(&ring);
 
         /* Poll for completions */
         int remaining_completions = num_pipes * 2;
```

**Comment by axboe** [comment]
Queue size bump is just for larger number of pipes, ran out at 4000 pipes since we need 2x that. For smaller number of pipes, the current 5000 is fine and won't make any difference.

**Comment by Qix-** [comment]
@axboe mind sharing some system details? Is this a system one might easily recreate?

**Comment by axboe** [comment]
Definitely, I think I mentioned some details higher up somewhere, it's the same system. Just something I put together as a compile/test box at the office:

- AMD Ryzen Threadripper 3970X 32-Core Processor, TRX40 board of some sort
- 32G of RAM

Kernel config attached for the non-mitigation runs.

[amd-config.gz](https://github.com/axboe/liburing/files/5975925/amd-config.gz)

**Comment by lano1106** [comment]
Hi,

can someone here tell me if io_uring could be beneficial for the system latency by using it inside a client have 1-2 TCP connections with a server? (It is an application very sensitive to latency where every nanosecond is precious...)

I do have such client that is using libev which has untested io_uring support. If there is a chance that it could help, this might make me interested in participating into the benchmarking effort for that particular scenario.

thank you,

**Comment by alexhultman** [comment]
Well that's what we are trying to figure out. **Theory** says yes, but my two old Intel CPUs say no. New AMD Threadrippers say yes. Conclusion pending..

**Comment by alexhultman** [comment]
![Screenshot from 2021-02-14 17-59-45](https://user-images.githubusercontent.com/3397140/107883329-45f2b700-6ee6-11eb-9b8a-385097e75a39.png)


I applied Jens's patch and ran at 4000 pipes and in this case I get a clear win for io_uring with mitigations off. But for smaller numbers epoll still wins. At least this is the first time I see anything go in favor of io_uring on my machine so this is good news but still, I would like to see 100 pipes go to io_uring.

I will keep up with 5.12 when it stabilizes and run with smaller numbers and hopefully some day we can have io_uring clean the table in all categories.

**Comment by axboe** [comment]
100 pipes is the same runtime here for epoll and i_uring with the 5.11 tree, and about 17% faster with the 5.12 tree for io_uring. Branch should be fully stable if you want to give it a go, I'd just recommend merging it into the current -git tree as it's based on an earlier 5.11-rc.

5.12 branch:

```
# taskset -c 0 ./io_uring 100
Pipes: 100, BUF_SIZE=32
Time: 0.248905

# taskset -c 0 ./epoll 100
Pipes: 100, BUF_SIZE=32
Time: 0.301584
```

io_uring has slightly higher setup/teardown cost, but doesn't seem to impact this test (ie running 10x the amount of loops shows same percentage win).

**Comment by alexhultman** [comment]
I've only picked tarballs from kernel.org before. What exact git branches should I do what with? I should have time to test this 5.12 today if I can just figure out what to clone from a command line perspective.

**Comment by isilence** [comment]
https://git.kernel.dk/cgit/linux-block/log/?h=for-5.12/io_uring

```
git clone:
git://git.kernel.dk/linux-block
or
https://git.kernel.dk/linux-block

branch: for-5.12/io_uring
```

Copy your config, e.g. from `/proc/config.gz`, compile (i.e. make -j10), and deploy.

Don't put it into a production environment, it's for-next, meaning not yet stabilised early changes, so may contain bugs/problems.

**Comment by alexhultman** [comment]
With that 5.12-branch there's definitive improvements, but for the 100 case there's still not any real win for io_uring but not a loss either. For the 200 case there is a significant win for io_uring now, even on my computer. So... still some work to be done but things are moving in the right direction for sure.

![Screenshot from 2021-02-15 12-15-17](https://user-images.githubusercontent.com/3397140/107945650-d6cd9f00-6f87-11eb-8f00-27ea0eff97b7.png)
![Screenshot from 2021-02-15 12-15-49](https://user-images.githubusercontent.com/3397140/107945652-d7663580-6f87-11eb-9c6f-6547c67e20b0.png)

**Comment by alexhultman** [comment]
Interesting.. io_uring is no longer **slower** even for small cases...

![Screenshot from 2021-02-15 12-23-44](https://user-images.githubusercontent.com/3397140/107946335-ce299880-6f88-11eb-96ab-33dea0415528.png)

**Comment by isilence** [comment]
@alexhultman, you may be seeing something else but io_uring or epoll pure overhead. I'd recommend to at least do more iterations, under 1s runs are not really great for benchmarking, ~1 minute would be nicer (to warm up caches, stabilie clock rates, etc.). Or even to profile it, e.g. `perf top`, because Jens does see a win, and looks it's pretty much reproducible, so it's not clear what's wrong with your setup. Also, don't forget about taskset or cset to exclude crazy task migration problems.

BTW, 5.12 is with the mentioned ~40% cut from the io_uring specific overhead. There is always space for improvements but for your case the io_uring  part should very slim already. It's a good question how much of it is in the pipe code.

**Comment by alexhultman** [comment]
Right, I made it go 10x the iterations and epoll still leads by a small bit:

```
alexhultman@clr-9c9c4beff725453db9688b589234d573~/io_uring_epoll_benchmark $ sudo ./io_uring 50
Pipes: 50, BUF_SIZE=32
Time: 3.001280
alexhultman@clr-9c9c4beff725453db9688b589234d573~/io_uring_epoll_benchmark $ sudo ./io_uring 50
Pipes: 50, BUF_SIZE=32
Time: 2.993102
alexhultman@clr-9c9c4beff725453db9688b589234d573~/io_uring_epoll_benchmark $ sudo ./io_uring 50
Pipes: 50, BUF_SIZE=32
Time: 2.993821
alexhultman@clr-9c9c4beff725453db9688b589234d573~/io_uring_epoll_benchmark $ sudo ./epoll 50
Pipes: 50, BUF_SIZE=32
Time: 2.953081
alexhultman@clr-9c9c4beff725453db9688b589234d573~/io_uring_epoll_benchmark $ sudo ./epoll 50
Pipes: 50, BUF_SIZE=32
Time: 2.953307
alexhultman@clr-9c9c4beff725453db9688b589234d573~/io_uring_epoll_benchmark $ sudo ./epoll 50
Pipes: 50, BUF_SIZE=32
Time: 2.939106
alexhultman@clr-9c9c4beff725453db9688b589234d573~/io_uring_epoll_benchmark $ 
```

**Comment by alexhultman** [comment]
I have an Asus N56Vz https://www.asus.com/us/Laptops/N56VZ/ and I've used this for like 7 years. There is absolutely nothing wrong with it and I have tested on other machines as well. I do see the gains in 5.12 - there is a huge gain for io_uring for me too, but tests still put epoll in better position doing smaller cases:

```
alexhultman@clr-9c9c4beff725453db9688b589234d573~/io_uring_epoll_benchmark $ sudo taskset 1 ./io_uring 50
Pipes: 50, BUF_SIZE=32
Time: 60.225375
alexhultman@clr-9c9c4beff725453db9688b589234d573~/io_uring_epoll_benchmark $ sudo taskset 1 ./epoll 50
Pipes: 50, BUF_SIZE=32
Time: 58.814421
alexhultman@clr-9c9c4beff725453db9688b589234d573~/io_uring_epoll_benchmark $ 
```

**Comment by isilence** [comment]
It may be subtle. Mind to attach what `perf top` says for both io_uring and epoll tests?

**Comment by alexhultman** [comment]
I've never used perf top - can you tell me what info you want and how to get it? How should I run the tests?

**Comment by lnicola** [comment]
FWIW, I can confirm @alexhultman's poor results with `io_uring`, especially when limiting it to a single CPU (it doesn't finish in a reasonable amount of time).

```
$ ./io_uring 1000
Pipes: 1000
Time: 6.735009

$ ./epoll 1000
Pipes: 1000
Time: 3.771896
```

Tested with 1 KB buffers and @alexhultman's code, but with the single `io_uring_submit` patch from @axboe.

`perf` shows almost all time as being spent in `io_uring_peek_batch_cqe`, assuming the symbols are correct:

![image](https://user-images.githubusercontent.com/308347/137126450-e44a4c54-b26d-405f-8082-4e7831ee3ff2.png)

Linux 5.14.11, `liburing` 2.1

**Comment by isilence** [comment]
Weird, got an email notification with @alexhultman claiming that io_uring outperformed 

there is a email from

**Comment by isilence** [comment]
> FWIW, I can confirm @alexhultman's poor results with `io_uring`, especially when limiting it to a single CPU (it doesn't finish in a reasonable amount of time).
> 
> Tested with 1 KB buffers and @alexhultman's code, but with the single `io_uring_submit` patch from @axboe and a change to link `liburing` dynamically.

@lnicola, well, IIRC the test is crafted in such a way that everything completes inline, and so it shouldn't really wait on CQEs. Apparently it's not the case for you, and the test uses CQ busy polling instead of waiting, which is an horror.

**Comment by lnicola** [comment]
> Weird, got an email notification with @alexhultman claiming that io_uring slightly outperformed epoll on some setup, but can't find it here.

Yeah, they've removed the comment.

> Apparently it's not the case for you, and the test uses CQ busy polling instead of waiting, which is an horror.
@lnicola

It's a bit weird. I tried to replace `io_uring_peek_batch_cqe` with `io_uring_wait_cqe_nr` and it locks up my computer and/or doesn't finish (even with `wait_nr` set to 1).

**Comment by mwojterski** [comment]
@alexhultman would that mean the conclusions in README on https://github.com/alexhultman/io_uring_epoll_benchmark are no longer up to date?

**Comment by lnicola** [comment]
Firefox says stuff like "pipe error: Too many references" and "Channel error: cannot send/recv", then my system locks up completely, which may or may not be an OOM (OOM also locks up my computer, even with NVMe swap).

It also happens in the polling mode with a large number of pipes.

**Comment by NobodyXu** [comment]
> It's not spending a lot of time in copying and like I mentioned Raspberry Pi 4 has completely different results on the same very test.
> 
> [output.txt](https://github.com/axboe/liburing/files/7340873/output.txt) ![perf](https://user-images.githubusercontent.com/3397140/137201817-04624495-5e6b-48db-a2db-ba643d98ade4.png)

That could be because the program calls io_uring_submit in a loop?

I think io_uring_enter, the underlying syscall, can submit multiple IOs in one syscall and wait for them, so maybe io_uring_submit_and_wait is more fitted here?

**Comment by NobodyXu** [comment]
> > That could be because the program calls io_uring_submit in a loop?
> 
> We tackled that a year ago already. If you read the documentation you can see what it does. It calls io_uring_submit once. We are just walking in a circle saying the same things we said a year ago now. Please look at this thread for history.

Sorry for that, it happened so long ago that I can’t remember the details.

**Comment by lnicola** [comment]
It does call submit in a loop, but fixing that doesn't make it significantly faster on my system.

**Comment by alexhultman** [comment]
I can recompile with debug symbols. These are very small copies, like we have talked about already, it's 32 bytes per pipe so overhead of copy is very likely not the bottleneck and again, if you look in this thread you can see that many people with AMD CPUs see big wins with io-uring and, again, my own raspberry pi is much faster with io uring. They all do the same copying, yet my shitty Intel machines don't perform any better with iouring

**Comment by isilence** [comment]
> I can recompile with debug symbols. These are very small copies, like we have talked about already, it's 32 bytes per pipe so overhead of copy is very likely not the bottleneck and again, if you look in this thread you can see that many people with AMD CPUs see big wins with io-uring and, again, my own raspberry pi is much faster with io uring. They all do the same copying, yet my shitty Intel machines don't perform any better with iouring

Ok, still curious to see relative overheads. If you're willing to recompile, there is a list of kernel options to check
(from https://www.brendangregg.com/perf.html)


```
# for perf_events:
CONFIG_PERF_EVENTS=y
# for stack traces:
CONFIG_FRAME_POINTER=y
# kernel symbols:
CONFIG_KALLSYMS=y
# tracepoints:
CONFIG_TRACEPOINTS=y
# kernel function trace:
CONFIG_FTRACE=y
# kernel-level dynamic tracing:
CONFIG_KPROBES=y
CONFIG_KPROBE_EVENTS=y
# user-level dynamic tracing:
CONFIG_UPROBES=y
CONFIG_UPROBE_EVENTS=y
# full kernel debug info:
CONFIG_DEBUG_INFO=y
# kernel lock tracing:
CONFIG_LOCKDEP=y
# kernel lock tracing:
CONFIG_LOCK_STAT=y
# kernel dynamic tracepoint variables:
CONFIG_DEBUG_INFO=y
```

**Comment by vcaputo** [comment]
@alexhultman The io_uring variant of your benchmark completes consistently faster in my tests when I change the `continue;` style CQ polling to instead wait for a CQE to arrive via `io_uring_wait_cqe()` when the CQ isn't ready:
```
             if (completions == 0) {
-                continue;
+
+               if (io_uring_wait_cqe(&ring, cqes) < 0) {
+                       printf("error waiting for any completions\n");
+                       return 0;
+               }
+               completions = 1;
             }
```

@axboe Is this style of brute-force CQ polling *supposed* to work well?  In my tests, especially when I add `taskset -c 0`, the io_uring test takes *much* longer without the `io_uring_wait_cqe()`-when-unready.  At a glance, at the very least there's the io_uring-created sqp sibling thread that also requires User CPU time, which the spinning is contending with.  If there's supposed to be some magic in how the sqp thread is created to make it complement the process-wide `taskset -c 0` nicely, and not fight with this style of polling, it doesn't seem to be working correctly here (5.14.5-arch1-1, i7-3520M).

I'm on a somewhat older Intel CPU, and adding the wait as noted above makes the io_uring test consistently faster than epoll.  From where I'm sitting right now, the io_uring benchmark seems to simply be incorrectly implemented.

**Comment by alexhultman** [comment]
@vcaputo The benchmark started like that, io_uring_submit_and_wait and io_uring_wait_cqe was how the original version of the benchmark ran. At that point (Linux 5.8) it was still not faster than epoll so the benchmark changed to make use of fixed files, fixed buffers and polling. That was faster so that solution remained. I can re-check with io_uring_wait_cqe on my Linux 5.16. Thanks for testing and reporting.

**Comment by alexhultman** [comment]
@vcaputo That changes doesn't do any difference for me. It never even runs that path.

**Comment by isilence** [comment]
> PS: there's something funny about the test, it sometimes fails to retrieve a SQE, even though the ring seems to be appropriately-sized.

@lnicola, I just noticed that it enables SQPOLL, and the program doesn't handle it right, though should work fine with normal rings. Anyone benchmarking it should disable SQPOLL unless willing to trink appropriately and invest more time in comparison. It's not magic, it can degrade performance and add variability to results.

**Comment by rbernon** [comment]
Hi, I've also been playing with io_uring latetly, trying to match epoll
performance in an IPC use case with pipes, one server and n clients, and
a simple request-reply protocol.

I modified the benchmark mentionned here and elsewhere for my tests, see
the `*ipc*.c` sources in https://github.com/rbernon/io_uring-echo-server

What I could see is that epoll currently always beats io_uring, with any
number of clients. My use of io_uring is pretty simple, I used the same
base as the existing code with fixed fds and buffers (although that did
not make a difference), and I'm submitting write-read linked sqes, with
the cqe_skip_success flag set on the write.

I've made some gpuvis traces, which I think are intersting to understand
the reasons. I used a build of the linux kernel from git, with the
patches from [1] on top (I verified that there's no cqe returned for the
write completion.

With epoll, with 1 client thread, each thread wakes the other one when
it has completed its write, and they alternate very quickly.

![epoll-1](https://user-images.githubusercontent.com/4134719/144759544-1098bff5-b2a8-4322-92c6-ac5bb5cc3683.png)

With 4 client threads, the server is fully busy (in red), and never gets
interrupted, serving requests one after another, and the processing time
is still very small and roughtly the same as with one client.

![epoll-4](https://user-images.githubusercontent.com/4134719/144759548-7f94cb28-2424-4403-a296-b336356d58ec.png)

With io_uring, however things are quite different, even with 1 client.
For a start, there's a kernel worker involved, so a third thread. Then,
there seems to be at least one spurious wakeup of the server.

![io_uring-1](https://user-images.githubusercontent.com/4134719/144759554-a8c30269-876a-451a-880a-bde167a70624.png)

I would expect that the client write would only wake up the worker,
copying the data and waking up the server, which would process it,
submit the next sqes, waking the worker and starting to wait for the
next one. I verified that the only cqe that the server receives is for
the read completion, but maybe it resumes from io_uring_submit_and_wait
for nothing once.

The number of spurious wakeups doesn't see to depend on the number of
clients, as can be seen here, with 2 clients. However, there's an
additional worker for each client, it doesn't really seem necessary?

![io_uring-2](https://user-images.githubusercontent.com/4134719/144759561-8e5ba6d7-d553-476e-9d9d-82b3bc1da928.png)

About the io_uring workers, I think there's something fishy, as whenever
I stopped the measuring, all went crazy for a bit, as can be seen
here (note that the scale is much larger, and every small block is also
an io_uring worker working).

![io_uring-crazy](https://user-images.githubusercontent.com/4134719/144759572-11b6f56d-cab3-4a89-89c4-ff86a66b04b6.png)

[1] https://lore.kernel.org/all/cover.1636559119.git.asml.silence@gmail.com

**Comment by Qix-** [comment]
@rbernon just out of curiosity, which tool is that?

**Comment by rbernon** [comment]
https://github.com/mikesart/gpuvis

**Comment by axboe** [comment]
If you see any iou-wrk workers, then something isn't working right. That's the slower async path, and should not get hit unless IOSQE_ASYNC is being explicitly set. What code is being run?

**Comment by axboe** [comment]
Took a quick look at that server, and it looks like it has two modes:

1) Use IOSQE_ASYNC, or
2) Use an explicit poll

Neither are great solutions, the most performant will generally be to just issue the request and have the internal poll handle it. Otherwise you're just trading an epoll readiness based model for the ditto on io_uring, which doesn't make a lot of sense. It can be useful for converting existing code, but for new code it isn't really advisable.

**Comment by rbernon** [comment]
> If you see any iou-wrk workers, then something isn't working right. That's the slower async path, and should not get hit unless IOSQE_ASYNC is being explicitly set. What code is being run?

Interesting, I added the flag on the read sqe because it's not expected to succeed immediately after a write from the server (the clients are not generally spamming requested). It didn't seem to make a difference at first, but I'll try without it.

**Comment by rbernon** [comment]
> Took a quick look at that server, and it looks like it has two modes:
> 
>     1. Use IOSQE_ASYNC, or
> 
>     2. Use an explicit poll
> 
> 
> Neither are great solutions, the most performant will generally be to just issue the request and have the internal poll handle it. Otherwise you're just trading an epoll readiness based model for the ditto on io_uring, which doesn't make a lot of sense. It can be useful for converting existing code, but for new code it isn't really advisable.

FWIW the server code is https://github.com/rbernon/io_uring-echo-server/blob/master/io_uring_ipc_server.c, and it only used the IOSQE_ASYNC flag on the read sqe.

I tried again without this flag, and with IOSQE_CQE_SKIP_SUCCESS, and now the spurious wakeup is gone but there's still iou workers involved, as well as the worker frenzy when exiting the test.

Here with 1 client:
![io_uring-1](https://user-images.githubusercontent.com/4134719/144814297-08fb9a6b-bf88-4f12-a807-d14e477685b4.png)

And with 2:
![io_uring-2](https://user-images.githubusercontent.com/4134719/144814334-e7f53d0b-b3b2-4e7d-87a3-beb194487452.png)

With 4 clients, something weird started to happen, with periods of "normal" latency and periods of very high latency. Looking at gpuvis I can see that the "normal" latency periods involve server, clients, and workers talking to each other:

![io_uring-4-normal](https://user-images.githubusercontent.com/4134719/144815119-595ca4e0-cacb-419a-8b92-057966b3473f.png)

But after a while there's only the server left, seemingly doing nothing?
![io_uring-4-weird](https://user-images.githubusercontent.com/4134719/144815158-a1854552-49f9-4f7c-a060-312426861ce2.png)

Plus a huge frenzy at the end, but kind of expected.

Unrelated note: by default gpuvis uses a millisecond resolution, I modified it to round the timestamps to us because I wanted to be able to see the latency more accurately.

**Comment by isilence** [comment]
@rbernon, pipes, right? The pipes internals don't support any sane nowait behaviour so we have to force any I/O against them to io-wq (slow path). It's just a one line on io_uring side to change that, but then submission may end up waiting for I/O potentially unboundedly. Hopefully, one day we'll push good enough for a change in the pipe code.

fwiw, the benchmark up in the thread doesn't go through io-wq only because O_NONBLOCK and there is special case in io_uring for that.

Don't know what that "frenzy" with lots of rescheduling exactly is, though that's interesting.

**Comment by avikivity** [comment]
Can't we register a poll on the pipe, and issue the write from the poll callback?

The same holds for other socket-like files that support poll() but aren't sockets. They can be handled in the same way as recv.

**Comment by avikivity** [comment]
Now I'm not sure that recv completes without a workqueue.

Looking at the code, it ends up in io_apoll_task_func, which punts via io_req_task_submit. Is recv indeed using workqueue?

**Comment by avikivity** [comment]
I'm probably lost in the maze.

**Comment by axboe** [comment]
We do use poll for any file type that supports it, but since we can't issue a nonblocking read attempt on a pipe, it still has to be done from a worker. What needs to happen here is just converting pipe from using struct file_operations->read to ->read_iter. The solution is known, and patches do exist.

This isn't specific to pipes, but obviously they are one of the more important file types. Thankfully most files use ->read_iter and ->write_iter these days.

**Comment by axboe** [comment]
> Looking at the code, it ends up in io_apoll_task_func, which punts via io_req_task_submit. Is recv indeed using workqueue?

It is not, this is task_work. That's different from thread offload. For the latter, look for io_queue_async_work().

**Comment by Cagoh** [comment]
I'm trying to fork() the io_uring simulation by creating parent and child process, each process have its own rings, parent for writing to the pipe, child for reading from the pipe, however I stumbled upon some error when the reader side read an empty pipe and instead of waiting for content in the pipe (if therers no error the simulation indeed run faaster), it update the CQE Async task failed. To overcome the problem, I didn't set the O_NONBLOCK in the pipe2(), this makes the reader to wait until theres content in the pipes before unblocking. But on the downside, it will call fcntl() with call syscall and the performance of the simulation is slower than the original simulation file. I wanted to know how to set reader side of pipe to be ready without syscalls

**Comment by ywave620** [comment]
This is a general question. Does replacing epoll with io_uring reduce the CPU usage for a given workload, said an nginx like http gateway? Reducing CPU usage in our production environment is the primary task rather than achieving a higher throughput or RPS in the benchmark environment.

This question comes to me because, IMO, io_uring makes socket IO purely asynchronous, but the data copying between user and kernel is preserved, assuming zero copy(IORING_OP_SEND_ZC, IORING_OP_SENDMSG_ZC) is not used.
